### ISSUE TITLE
[BUG] Black main screen 

### ISSUE THREAD

**Issue by tdelc** [daid/EmptyEpsilon#649]
There is a bug in the main screen : in some situations, the viewport disappear and all the screen is black.
How to do it : 
- create more than one playerspaceship. The problem don't appear with the first person
- choose the main screen for a playerspaceship (not the first) and choose first person view ("F")
- destroy all object in front of the playerspaceship
- the screen go black

I think that the viewport does not appear if there is not object to show. So, with the first person view, it can be nothing to show. The strange thing is that it does not the case for the first player.

Do you have some advice to correct it ?

**Comment by aBlueShadow** [comment]
Sounds similar to #636

Have you checked if it is really the order of the players, or specific computers (and therefore different graphic cards)?

**Comment by tdelc** [comment]
I test on several computer and it is not the same case.
In fact, if you create a ship into the ship view, the screen appear.

And Eventually, I think that this problem occur for all the playerspaceship

**Comment by aBlueShadow** [comment]
Fascinating.Yes, I can confirm that issue. Interestingly, when I tested it doesn't matter whether it is the first ship or not. Also, it is not needed to first log into the mainscreen and destroy objects afterwards. If there is nothing to look at in the first place, it will already be black.

> and it is not the same case.
I am still not sure about that, both issues are extremely similar, and they both seem to be present since 20190119. It could be that @PlanetariumWSD experienced a special case of that issue, where the mentioned headings were the ones where no object could be seen on their scenario.
Or there maybe they are two manifestation of the same graphical issues.

@daid The issue also appears on 20190119, but not on old releases. It seems to be graphics-card-independent and also occurs on both SFML 2.4 and 2.5. So it might have been introduced by SFML, but seems not to be solvable by swapping libraries.
That leads to another question: @PlanetariumWSD, you mention that your black screen issue disappeared after you recompiled EE with SFML2.5.1. Can you check if the issue that is described here is present on your version?

**Comment by PlanetariumWSD** [comment]
I can confirm this black screen first person camera (main screen and ship window) problem in SFML>2.3x builds as still present.   :-(
Exactly as described [here](https://github.com/daid/EmptyEpsilon/issues/636#issue-422836650).
Bug report #649 and #636 are the same issue.
Previous SFML 2.3 version works perfectly.
The flag added in merge #645 fixed only part of the problem. ( disabling warp/jump post processors).
Now that I think about it, I had seen the black screen issue but thought only on windowed (non-standard non power-of-2 aspect ratios).. I thought I'd just leave aspect ratio alone on our PCs BUT this is still a problem... as it's persistent on any aspect ratio... dare I say ARGHHHHhh...

**Comment by aBlueShadow** [comment]
Okay, I did some tests, and here are the results. It seems like the issue is one version older than expected, as I can reproduce it on 2018.11.16, but  not on 2018.09.06. 
With this knowledge, I tried to recompile 2018.09.06 using SFML 2.4.1. Still no issue!
So it seems, other than supsected, this issue is **not SFML-related**.

So the quastion is: were there some changes in terms of rendering/graphics between 2018.09.06. and 2018.11.16? (I know daid said there were no changes, but that was when we thought the bug appeared not until version 2019.01.19)

**Comment by aBlueShadow** [comment]
Found it!
I searched for rendering related PR in the timeframe in questions, but found only the sfml-related #591
However, part of this PR is 3f4524bc389be38fc03c35e8e3b64377db4e414e that interchanges two GL functions. By reverting that commit, the issue dissapeared.
Now, honestly I have no Idea what those functions and that commit do, so I won't do a reverting PR yet. But at least now we know a (potential) workaround.

@amir-arad what did that commit do exactly, and will reverting it cause other serious issues?

**Comment by PlanetariumWSD** [comment]
Cool! Confirming @aBlueShadow! Building with SFML 2.5.1 reverting from https://github.com/daid/EmptyEpsilon/commit/3f4524bc389be38fc03c35e8e3b64377db4e414e fixes both issues #649 and #636 but reintroduces it's intended fix for heading value font in first person camera modes from becoming white blocks as described https://github.com/daid/EmptyEpsilon/pull/578#issuecomment-431937133

**Comment by aBlueShadow** [comment]
ah, I see! Did not see the artifacts before as I built it with SFML 2.4.1

**Comment by amir-arad** [comment]
Sorry I missed being mentioned on this thread.
I wrote [3f4524b](https://github.com/daid/EmptyEpsilon/commit/3f4524bc389be38fc03c35e8e3b64377db4e414e) when looking for a fix for a specific SFML 2.5 compatibility issue.

I've just downloaded `EmptyEpsilon-2019.05.21-win32` and wasn't able to reproduce the error locally. 
The only other experience I have with EE not reproducing bugs on every station involves access violations that only happened on computers with Nvidia GPU (my station has no GPU other the one embedded in i5 gen6 ). namely nvoglv32.dll is in all the crash reports. 

could you write down the environment in which it reproduces? I would love to investigate further, but I need a faster feedback loop (local reproduction) to do that.

**Comment by aBlueShadow** [comment]
There are 2 conditions that have to be met for this bug to appear.
1) There have to be  nothing in the ships view
2) the view have to be first person (so either by pressing f on the mainscreen or using ship window)

Technically 2) is a subset of 1) because in third person, there is always at least one rendered object. 

If there is no object in the view, no skybox is shown.

So the easiest way to reproduce it is to create a completely empty scenerio, or to start "empty space" and quickly delete all objects in the GM screen.
Then spawn a ship, and either go in mainscreen and press "F", or start a ship window. The screen should then be black. That issue occurs on different graphic card manufacturers, on linux and also on the windows version (at least in WINE)

**Comment by aBlueShadow** [comment]
For reference:
```
-- Name: Pretty Empty
-- Description: 
-- Type: Basic

function init()

player = PlayerSpaceship():setFaction("Human Navy"):setTemplate("Atlantis") 

end 
function update(delta) 
end
```

Would it reproduce on you machine with this test scenario?

**Comment by amir-arad** [comment]
roger that @aBlueShadow . will re-run and report

**Comment by amir-arad** [comment]
after some analysis-paralysis, pragmatic solution seems to work. 
see #676

**Comment by daid** [comment]
Well, not setting up the projection matrix would screw up the rest of the rendering, so this might need to be moved outside the loop:
```
        glMatrixMode(GL_PROJECTION);
        glLoadIdentity();
        _glPerspective(camera_fov, rect.width/rect.height, 1.f, 25000.f * (n + 1));
        glMatrixMode(GL_MODELVIEW);
```

**Comment by amir-arad** [comment]
@daid the projection matrix is already set when the loop begins. 
I narrowed it down to resetting the texture `sf::Texture::bind(NULL);` by elimination.
